[testing framework] Add documentation (#33454)
hashicorp/terraform: website/docs/language/tests/index.mdx

## Integration vs Unit testing

By default, tests within Terraform create real infrastructure and can run assertions and validations against the created infrastructure. This can be imagined as integration testing, as the full Terraform stack is being executed and infrastructure being created.

It is possible to override this behaviour by updating the `command` attribute within a [`run`](#run-blocks) block (discussed below). By default, the each `run` block executes with `command = apply` causing Terraform to execute a complete `apply` operation against your configuration. Replacing this with `command = plan` will ensure Terraform will not create any new infrastructure for this `run` block, and test authors can validate any logical operations and custom conditions within their infrastructure in a manner analogous to unit testing.

## Syntax

Each Terraform test is contained within a test file. Test files are discovered by Terraform due to their file extension: `.tftest.hcl` or `.tftest.json`.

Each test file contains the following root level attributes and blocks:

- Zero to one [`variables`](#variables) block.
- Zero to many [`provider`](#providers) blocks.

> - One to many [`run`](#run-blocks) blocks.

Author: Done!

The `run` blocks are executed in order, simulating a series of Terraform commands being executed directly within the configuration directory. The order of the `variables` and `provider` blocks doesn't matter, all values within these blocks are processed once at the beginning of the test operation. A well laid out test file has the `variables` and `provider` blocks defined first, at the beginning of the file.

### Example

> The following example demonstrates a simple Terraform configuration that creates an AWS S3 bucket, using an input variable to modify the name, combined with a test that verifies the name of the S3 bucket is as expected.

Author: It now says that it is running a plan command in the test explanation I wrote for the previous comment. Do you think that is sufficient?

> 

Author: I've updated this to:

```
The following example demonstrates a simple Terraform configuration that creates an AWS S3 bucket, using an input variable to modify the name. We will create an example test file (below) that validates the buckets name is created as expected.
```

```hcl
# main.tf

provider "aws" {
    region = "eu-central-1"
}

variable "bucket_prefix" {
  type = string
}

resource "aws_s3_bucket" "bucket" {
  bucket = "${var.bucket_prefix}-bucket"
}

output "bucket_name" {
  value = aws_s3_bucket.bucket.bucket
}
```

> 

Author: I moved the explanation I had below the config to here.

```hcl
# valid_string_concat.tftest.hcl

variables {
  bucket_prefix = "test"
}

run "valid_string_concat" {

  command = plan

  assert {
    condition     = aws_s3_bucket.bucket.bucket == "test-bucket"
    error_message = "S3 bucket name did not match expected"
  }

}
```

The above test file runs a single Terraform plan command which creates the S3 bucket, and then validates the logic for calculating the name is correct by checking the actual name matches the expected name.

## Run blocks

> Each `run` block has the following fields and blocks:

Author: Is the following okay?

```
- An optional `command` attribute, which is either `apply` or `plan` and defaults to `apply`.
- An optional `plan_options` block, which contains:
  - An optional `mode` attribute, which is either `normal` or `refresh-only` and defaults to `normal`.
  - An optional boolean `refresh` attribute, which defaults to `true`.
  - An optional `replace` attribute, which contains a list of resource addresses referencing resources within the configuration under test.
  - An optional `target` attribute, which contains a list of resource addresses referencing resources within the configuration under test.
- An optional [`variables`](#variables) block.
- An optional [`module`](#modules) block.
- An optional [`providers`](#providers) attribute.
- Optional [`assert`](#assertions) blocks.
- An optional [`expect_failures`](#expecting-failures) attribute.
```

## Variables

You can provide values for [Input Variables](/terraform/language/values/variables) within your configuration directly from your test files.

The test file syntax supports `variables` blocks at both the root level and within `run` blocks. Variable values provided within the file will be passed into all `run` blocks within the file, and can be overridden for a particular `run` block by the variable values provided with that `run` block.

Continuing our [example](#example) from above:

```hcl
# variable_precedence.tftest.hcl

variables {
  bucket_prefix = "test"
}

run "uses_root_level_value" {

  command = plan

  assert {
    condition     = aws_s3_bucket.bucket.bucket == "test-bucket"
    error_message = "S3 bucket name did not match expected"
  }

}

run "overrides_root_level_value" {

  command = plan

  variables {
    bucket_prefix = "other"
  }

  assert {
    condition     = aws_s3_bucket.bucket.bucket == "other-bucket"
    error_message = "S3 bucket name did not match expected"
  }

}
```

> 

Author: Done!

By default, within each `run` block all defined providers will be made directly available. It is also possible to customize which providers are made available within a given `run` block using a `providers` attribute. The behaviour and syntax for this block matches the behaviour of [providers meta-argument](/terraform/language/meta-arguments/module-providers).

If no provider configuration is provided within a testing file, Terraform will attempt to initialize any providers within the configuration using their default settings. For example, any environment variables aimed at configuring providers will still be available and will be used by Terraform to create default providers.

Extending the previous [example](#example) to allow our tests, instead of the configuration, to specify the region:

```hcl
# main.tf

terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }
  }
}

variable "bucket_prefix" {
  type = string
}

resource "aws_s3_bucket" "bucket" {
  bucket = "${var.bucket_prefix}-bucket"
}

output "bucket_name" {
  value = aws_s3_bucket.bucket.bucket
}
```

> 

Author: Done!

# customised_providers.tftest.hcl

provider "aws" {
  region = "us-east-1"
}

provider "aws" {
  alias  = "secondary"
  region = "eu-central-1"
}

run "providers" {

  command = plan

  assert {
    condition     = aws_s3_bucket.primary_bucket.bucket == "test-primary"
    error_message = "invalid value for primary S3 bucket"
  }

  assert {
    condition     = aws_s3_bucket.secondary_bucket.bucket == "test-secondary"
    error_message = "invalid value for secondary S3 bucket"
  }
}
```

It is also possible to define specific providers you want to use in specific `run` blocks:

```hcl
# main.tf

terraform {
  required_providers {
    aws = {
      source                = "hashicorp/aws"
      configuration_aliases = [aws.secondary]
    }
  }
}

data "aws_region" "primary" {}

data "aws_region" "secondary" {
  provider = aws.secondary
}

variable "bucket_prefix" {
  default = "test"
  type    = string
}

resource "aws_s3_bucket" "primary_bucket" {
  bucket = "${var.bucket_prefix}-${data.aws_region.primary.name}-primary"
}

resource "aws_s3_bucket" "secondary_bucket" {
  provider = aws.secondary
  bucket   = "${var.bucket_prefix}-${data.aws_region.secondary.name}-secondary"
}
```

> 

Author: Done!

```hcl

> provider "aws" {

Author: yep, good shout!

  }
}

run "providers" {

  command = plan

  providers = {
    aws           = aws
    aws.secondary = aws.tertiary
  }

  assert {
    condition     = aws_s3_bucket.primary_bucket.bucket == "test-us-east-1-primary"
    error_message = "invalid value for primary S3 bucket"
  }

  assert {
    condition     = aws_s3_bucket.secondary_bucket.bucket == "test-eu-west-2-secondary"
    error_message = "invalid value for secondary S3 bucket"
  }
}
```

> **Note:** When running tests with `command = apply` switching providers between `run` blocks can result in failed operations and tests as resources created by one provider definition will be unusable when modified by a second.

## Modules

You can also modify the module that a given `run` block will execute.

By default, Terraform will execute the given command against the configuration under test for each `run` block. The configuration under test is the configuration within the directory the `terraform test` command was executed from (or the directory pointed to by `-chdir` argument). Each `run` block also allows the user to change the targeted configuration using the `module` block.

> Compared with the traditional [`module` block](/terraform/language/modules/syntax), the `module` block within test files only supports the [`source`](/terraform/language/modules/syntax#source) attribute and the [`version`](/terraform/language/modules/syntax#version) attribute. The remaining attributes that would normally be supplied via the traditional `module` block are provided elsewhere within the `run` block.

Author: It would be the user.

How about this?

```

Unlike the traditional [`module` block](/terraform/language/modules/syntax), the `module` block within test files _only_ supports the [`source`](/terraform/language/modules/syntax#source) attribute and the [`version`](/terraform/language/modules/syntax#version) attribute. The remaining attributes that are typically supplied via the traditional `module` block should be supplied by the alternate attributes and blocks within the `run` block.
```

> **Note:** Terraform test files only support [local](/terraform/language/modules/sources#local-paths) and [registry](/terraform/language/modules/sources#terraform-registry) modules within the `source` attribute.

All other blocks and attributes within the `run` block are supported when executing an alternate module, with `assert` blocks executing against values from the alternate module. This is discussed more in [Modules State](#modules-state).

Here are two example use cases for the `modules` block within a Testing file:

1. A setup module to create necessary infrastructure required by the main configuration under test.
2. A loading module to load and validate secondary infrastructure (such as data sources) not created directly by the main configuration under test.

The following example demonstrates both of these use cases:

> - We have a module that will create and load several files into an already created S3 bucket.

Author: Done!

  # configuration.

  # ... updated variables ...

  # ... assertions ...
}

run "loader" {

  # This run block is now referencing our second alternate module so will create
  # our third and final state file. The other two state files are managing
  # resources from the main configuration and resources from the setup module.
  # We are getting a new state file for this run block as the loader module has
  # not previously been referenced by any run blocks.

  module {
    source = "./testing/loader"
  }
}

```

#### Modules Cleanup

At the conclusion of a test file, Terraform will attempt to destroy every resource created during the execution of that test file. When alternate modules are loaded, the order in which objects are destroyed is important. For example, in our first [Modules](#modules) example earlier we cannot destroy the resources created in the "setup" `run` block before the objects created in the "execute" `run` block, as S3 bucket created in the "setup" step cannot be destroyed while it contains objects.

Terraform will destroy resources in the following order, and this order is important as it may affect the structure of your testing files:

> 1. Destroy the resources held in the main state file first, so you should not create resources in alternate modules that depend on resources from your main configuration.

Reviewer: ```suggestion
1. Resources in the main state file. Do not create resources in alternate modules that depend on resources from your main configuration.
```

>     - Note that data sources can refer to objects in your main configuration, as Terraform doesn't have to destroy data sources.

Reviewer: ```suggestion
    - Data sources can refer to objects in your main configuration, because Terraform does not have to destroy data sources.
```

> 2. Destroy the resources created by alternate modules in `run` block reverse order.

Reviewer: ```suggestion
2. Resources created by alternate modules in reverse `run` block order.
```

>     - From our example, any resources created in the "verify" `run` block would be destroyed before resources created in the "setup" `run` block. Note, that in our example this doesn't particularly matter as our "verify" `run` block only loads a data source and creates no resources.

Reviewer: Can we move this to be a paragraph below? I think the thought stands well enough below the list, and it will leave the list looking smaller and more scannable.

Reviewer: Also can you link back up to the example?

> If you only use a single setup module as an alternate module, and it executes first, or you use no alternate modules, then the order of destruction will not affect you. Anything more complex may require careful consideration to make sure the destruction of created resources can complete automatically.

Reviewer: ```suggestion
If you use a single setup module as an alternate module, and it executes first, or you use no alternate modules, then the order of destruction does not affect you. Anything more complex may require careful consideration to make sure the destruction of resources can complete automatically.
```

> ## Expecting Failures

Reviewer: ```suggestion
## Expecting failures
```

> By default, if any [Custom Conditions](/terraform/language/expressions/custom-conditions), including `check` block assertions, fail during the execution of a Terraform test file then the overall command will report the test as a failure. It is a common testing paradigm, however, to want to test failure cases. Terraform supports the `expect_failures` attribute for this use case.

Reviewer: Breaking it up for scanability:
```suggestion
By default, if any [Custom Conditions](/terraform/language/expressions/custom-conditions), including `check` block assertions, fail during the execution of a Terraform test file then the overall command reports the test as a failure. 

However, it is a common testing paradigm to want to test failure cases. Terraform supports the `expect_failures` attribute for this use case.
```

> In each `run` block the `expect_failures` attribute can provide a list of checkable objects (resources, data sources, check blocks, input variables, and outputs) that should fail their custom conditions. The test will then pass overall if these checkable objects report an issue, while the test will fail overall if they do not report an issue.

Reviewer: ```suggestion
In each `run` block the `expect_failures` attribute can provide a list of checkable objects (resources, data sources, check blocks, input variables, and outputs) that should fail their custom conditions. The test passes if the checkable objects you specify report an issue, and the test fails overall if they do not.
```

> You can still write assertions alongside an `expect_failures` block, but you should be mindful that all custom conditions, except check block assertions, halt the execution of Terraform. This still applies during test execution, so your assertions should only consider values that you are sure will be computed before the checkable object is due to fail. This can be managed via references or the `depends_on` meta-argument within your main configuration.

Reviewer: ```suggestion
You can still write assertions alongside an `expect_failures` block, but you should be mindful that all custom conditions, except check block assertions, halt the execution of Terraform. This still applies during test execution, so your assertions should only consider values that you are sure will be computed before the checkable object is due to fail. You can manage this using references, or the `depends_on` meta-argument within your main configuration.
```

> This also means that, with the exception of `check` blocks, only a single checkable object can be reliably included. We support a list of checkable objects within the `expect_failures` attribute purely for `check` blocks.

Reviewer: ```suggestion
This also means that, with the exception of `check` blocks, you can only reliable include a single checkable object. We support a list of checkable objects within the `expect_failures` attribute purely for `check` blocks.
```

> A quick example here demonstrates testing the `validation` block on an input variable.

Reviewer: ```suggestion
A quick example below demonstrates testing the `validation` block on an input variable.
```

```hcl
# main.tf

variable "input" {
  type = number

  validation {
    condition = var.input % 2 == 0
    error_message = "must by even number"
  }
}
```

> 

Reviewer: Add a sentence saying this is the test file

```hcl
# input_validation.tftest.hcl

variables {
  input = 0
}

run "zero" {
  # The variable defined above is even, so we expect the validation to pass.

  command = plan
}

run "one" {
  # This time we set the variable is odd, so we expect the validation to fail.

  command = plan

  variables {
    input = 1
  }

  expect_failures = [
    var.input,
  ]
}
```

> >> **Note**: Failures are only expected in the operation specified by the `command` attribute of the `run` block.

Reviewer: ```suggestion
>> **Note**: Terraform only expects failures in the operation specified by the `command` attribute of the `run` block.
```

> This means you should be careful when using `expect_failures` in `run` blocks with `command = apply`. A `run` block with `command = apply` that expects a custom condition failure will fail overall if that custom condition fails during the plan.

Reviewer: Starting the sentence with "this means" feels weird here, are you building off the note above? Could we do this instead?

```suggestion
Be careful when using `expect_failures` in `run` blocks with `command = apply`. A `run` block with `command = apply` that expects a custom condition failure will fail overall if that custom condition fails during the plan.
```

> This is logically consistent, as the `run` block is expecting to be able to run an apply operation but cannot as the plan failed. It is also potentially confusing, as you will see the failure reported in the diagnostics as the reason the test failed even though that failure was marked as being expected.

Reviewer: ```suggestion
This is logically consistent, as the `run` block is expecting to be able to run an apply operation but can not because the plan failed. It is also potentially confusing, as you will see the failure in the diagnostics as the reason the test failed, even though that failure was marked as being expected.
```

> There are instances when a custom condition will not be executed during the planning stage as it is relying on computed attributes only available after a referenced resource has been created. In these cases an `expect_failures` block alongside a `command = apply` attribute and value would be expected and acceptable. However, in most cases you should ensure that `expect_failures` are used only alongside `command = plan` operations.

Reviewer: ```suggestion
There are instances when Terraform does not execute a custom condition during the planning stage, because that condition is relying on computed attributes that are only available after Terraform creates the referenced resource. In these cases, you could use an `expect_failures` block alongside a `command = apply` attribute and value. However, in most cases we recommend only using `expect_failures` alongside `command = plan` operations.
```

>> **Note**: Expected failures only apply to user-defined custom conditions.

> This means that any other kind of failure in a specified checkable object will still result in the overall test failing. For example, a variable that expects a boolean value as input will fail the surrounding test if the wrong kind of value is provided even if that variable is included in an `expect_failures` attribute.

Reviewer: ```suggestion
Other kinds of failure _besides_ the specified expected failures in the checkable object still result in the overall test failing. For example, a variable that expects a boolean value as input fails the surrounding test if Terraform provides the wrong kind of value, even if that variable is included in an `expect_failures` attribute.
```

> The `expect_failures` attribute is included to allow authors to test their configuration and any logic defined within. A type mismatch, as in the previous example, is not something Terraform authors should have to worry about testing as Terraform itself will handle enforce type constraints. As such, you can only `expect_failures` in custom conditions.

Reviewer: ```suggestion
The `expect_failures` attribute allows authors to test their configuration and any logic defined within. A type mismatch, as in the previous example, is not something Terraform authors should have to worry about testing as Terraform itself will handle enforce type constraints. As such, you can only `expect_failures` in custom conditions.
```
